### ISSUE TITLE
V7 support for "old" FahCores (-forceasm)

### ISSUE THREAD

**Issue by jcoffland** [FoldingAtHome/fah-client-pub#115]
**Reported by @bb30994 on 26 Jul 2010 02:22 UTC**
Whenever I stop fahclient, the log ends (more or less) like this:
''01:46:40:WARNING: Console control signal 0 on PID 2668
01:46:40:Exiting, please wait. . .
01:46:41:Waiting on unit 01
01:47:41:WARNING: Killing unit 01
01:47:41:ERROR: Exception: 0: Core still running
01:47:41:Clean exit''

As far as the cores are concerned, this is NOT a clean exit.  FahCore_78 will restart with optimizations disabled, which means that -forceasm will ALWAYS be required when configuring a uniprocessor client.

I beleive your method of stopping the cores may be logical but apparently it's not the same method used by the previous client and universally understood by the FahCore's (except maybe b4).  

I'm only guessing here, but you are probably sending the appropriate signal to the core, which is being ignored ... until you kill the core.  I suspect that the Cores expect the client to shut itself down and they proceed to shut themselves down once the client is no longer present.  Whether my guess is correct or not, please study the shutdown sequence as performed by V3/v4/v5/v6 and FahCore_78 or any other GROMACS based core with assembly optimizations.

Do you intend to fix all of the existing FahCores to work with the new method?

**Comment by jcoffland** [comment]
**Comment by @jcoffland on 26 Jul 2010 21:46 UTC**
Your guesses are correct.  There are two problems here.  #1 The client log file says '''Clean exit''' but the core log file says '''improper shutdown''' when restarted.  #2 The '''Core still running''' exception.

I just made some changes to the code to fix #2.

In my opinion, it is a mistake that the cores do not provide away for the client to properly shut them down.  This is not just a problem with client shutdown but also with core pausing.  Some cores support a pause signal.  Most do not.  So we have to terminate the process to pause most cores.  For that reason '''improper shutdown''' is unavoidable.  Even though cores say this they have to be able to handle hard termination so it's not really a problem.

I also want to make sure the cores are really shutdown before the client exits.  One good reason for this is that it avoids the possibility of running two instances of the same core on the same WU at the same time.

So, although I agree with you that #1 is not pretty that is just the way it works for now.

I've thought about starting up a dummy process and using this as the core "life line".  I could then kill that process and the cores would get the signal to shutdown.  However, that is somewhat complicated and potentially one extra process per running core might look worse to the user than the '''improper shutdown''' message.

And to answer your other question.  No I don't plan to fix all existing cores right now but eventually I would like them to start listing to the console control event and shutdown nicely.

**Comment by jcoffland** [comment]
**Comment by @jcoffland on 10 Sep 2010 04:22 UTC**
Fixed in v7.0.13.  Please reopen if you see this error again.

**Comment by jcoffland** [comment]
**Comment by @bb30994 on 28 Feb 2011 00:59 UTC**
You fixed half of the original problem but not the other half.

If V7 is going to be released while certain FahCores (that use lifelines) are still supported, there needs to be some kind of change to V7's "killing" methodology.  Either that all of those cores need to be updated to work with V7 by accepting a gentle shutdown request or simply by assuming -forceasm.

Many Uniprocessor clients will be run by Novices who will not want to be messing around with <extra-core-args ... />

See http://foldingforum.org/viewtopic.php?f=65&t=17742

**Comment by jcoffland** [comment]
**Comment by @jcoffland on 28 Feb 2011 19:51 UTC**
This looks like a ticket hijacking to me.  I know you mentioned this topic when submitting the bug but this ticket was about cores not shutting down correctly.  I've created a new ticket #563.

As a rule of thumb please consider the ticket title to be the authoritative subject of the ticket regardless of what the ticket comments say.
